### PR TITLE
feat(Dialog): Add draggable dialog and rename existing Modal

### PR DESCRIPTION
**Pull request by elephantcatdog**
modal.scss
- reconfigured modal names and split the elements of the modal into different mixins
- `isolation: isolate` added to separate z-index of the modal components from the z-index of all components (line 10)
- `inset: 0` replaces `top: 0` and `left: 0` (lines 12 & 73)
- the visibility and opacity transitions for `.iui-modal-full-page` is inside `@media (prefers-reduced-motion: no-preference)` so that the full page modal will fade nicely by following the transition rules for standard modals (lines 26-30)
- `display: flex` and `flex-direction: column` added to only full page and resizable modals to avoid breaking modals in React (lines 76-77 & 100-101)
  - these lines should be added to `iui-modal` in React 2.0 (around line 50)
- `transform: translateY(100%)` is inside `@media (prefers-reduced-motion: no-preference)` so that the full page modal fades nicely when no motion is selected by the user (line 80)
modal.html
- wrote javascript for a draggable modal (lines 225-270)

collaboration between Jon and Leah

### CODE REVIEW

File: src/modal/classes.scss [hunk 1]
```diff
@@ -5,8 +5,24 @@
 .iui-modal {
   @include iui-modal;
 
-  &-full-page {
-    @include iui-modal-full-page;
+  &-backdrop {
```
Comment (gretanausedaite) [inline src/modal/classes.scss:8]: Why did you remove `&-full-page`?

File: src/modal/modal.scss [hunk 2]
```diff
@@ -153,3 +128,67 @@
...
+  box-sizing: border-box;
+  font-size: $iui-font-size-subheading;
+
+  @at-root .iui-modal-resizable & {
+    user-select: none;
+    font-size: $iui-font-size-leading;
+    padding: round($iui-baseline * 0.5) $iui-m;
+    cursor: grab;
+    @include themed {
+      background-color: t(iui-color-background-3);
```
Comment (gretanausedaite) [inline src/modal/modal.scss:154]: Is background color of the title bar in our standards? I'm not sure I like that simple and draggable/resizable modals have different title bars.

File: backstop/scenarios/modal.js [hunk 3]
```diff
@@ -6,13 +6,22 @@ module.exports = [
...
     actions: [click('#open-full-page-modal'), hover('body')],
     selectors: ['document'],
-    hideSelectors: ['body > *:not(.iui-modal)'],
+    hideSelectors: ['body > *:not(.iui-modal-backdrop)'],
+  }),
+
+  scenario('resizable-modal', {
+    // Hovering `body` to avoid random test failures
+    // when "Open modal" sometimes is and sometimes is not hovered.
+    actions: [click('#open-resizable-modal'), hover('body')],
```
Comment (gretanausedaite) [inline backstop/scenarios/modal.js:23]: Tests fail: `waiting for selector `#open-resizable-modal` failed: timeout 30000ms exceeded` 

Change this to `toggle-resizable-modal` or change line 70 in modal.html.

File: src/modal/classes.scss (hunk 1 above)
Comment (elephantcatdog) [reply src/modal/classes.scss:8]: The previous mixin was rolled into the mixin for `.iui-modal` along with resizable.

File: backstop/scenarios/modal.js (hunk 3 above)
Comment (elephantcatdog) [reply backstop/scenarios/modal.js:23]: Thanks!

File: src/modal/modal.scss (hunk 2 above)
Comment (elephantcatdog) [reply src/modal/modal.scss:154]: Jon did the visual work on this; I'll ping him to respond also. I know that the resizable modal's title bar is a different color to make it apparent that it's draggable (and resizable, I suppose?). If they look the same how can we signal to users that some are draggable/resizable?

Comment (FlyersPh9) [reply src/modal/modal.scss:154]: I was trying to match [what is currently implemented in AppUI](https://github.com/iTwin/iTwinUI-react/issues/614).  I think we need a bar area for the user to grab.  If you have any visual design ideas of how you'd like it to look, I'm open to investigating alternative designs.

File: src/modal/classes.scss (hunk 1 above)
Comment (FlyersPh9) [inline src/modal/classes.scss:8]: Let's change the structure of `classes.scss` to follow @veekeys [suggestion shared here](https://github.com/iTwin/iTwinUI/pull/598#discussion_r857320738).  Personally I like how `classes.scss` looks currently, but Vykintas' suggestion will definitely make it easier to search and find classes.

Comment (elephantcatdog) [reply src/modal/classes.scss]: Changed

Comment (elephantcatdog) [reply src/modal/classes.scss]: So we actually just worked on making `iui-modal-default`, `iui-modal-full-page`, and `iui-modal-resizable` into their own mixins also. We did this so that resizable modals won't need to have both of the classnames `iui-modal-default` and `iui-modal-resizable`, as `iui-modal-default` is now included into `iui-modal-resizable` in classes.scss.

File: src/modal/modal.scss [hunk 4]
```diff
@@ -2,154 +2,194 @@
...
+    // remove delay for entry animation so modal is instantly visible
     transition-delay: $iui-speed-instant;
   }
-  // #endregion Small Modal Animations for CSS
+}
 
-  > .iui-modal-dialog {
-    position: absolute;
+@mixin iui-modal {
+  @include iui-surface($border-radius: $iui-border-radius, $elevation: $iui-elevation-24);
```
Comment (bentleyvk) [inline src/modal/modal.scss:33]: Need to make Surface IE friendly:
image

File: src/modal/modal.scss [hunk 5]
```diff
@@ -2,154 +2,194 @@
...
-  // #endregion Full Page Modal Animations for CSS
+}
+
+@mixin iui-modal-resizable {
+  background-color: transparent;
+  pointer-events: none;
+  z-index: 998;
+
+  .iui-modal {
+    pointer-events: initial;
```
Comment (bentleyvk) [inline src/modal/modal.scss:95]: What it does?

File: src/modal/modal.scss (hunk 2 above)
Comment (gretanausedaite) [reply src/modal/modal.scss:155]: Darker bar area does not necessarily means that item is draggable.
We have this drag handle icon which could be included in title: https://itwin.github.io/iTwinUI-icons/#icons-drag-handle and I tried to think where (if) we could include it in bar area and it just doesn't fit.
So probably it's best to have darker title bar for draggable dialogs/modals. Especially when people are already using it like this.

File: src/modal/modal.scss (hunk 4 above)
Comment (elephantcatdog) [reply src/modal/modal.scss:33]: Reverted

**Comment by bentleyvk** [comment]
Why we now have dialog and modal with the same CSS?

File: src/dialog/dialog.scss [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+@mixin iui-dialog-resizable {
+  max-width: 100vw;
+  max-height: 100vh;
+  min-width: 380px;
+  min-height: 144px;
+  resize: both;
+}
+
+@mixin iui-dialog-animation {
```
Comment (bentleyvk) [inline src/dialog/dialog.scss:138]: Why separate mixin and class for animation? Just have it on the base dialog.

**Comment by FlyersPh9** [comment]
@bentleyvk @gretanausedaite:  Vykintas and I worked on this PR a little bit today because @elephantcatdog is out sick.  We think the HTML / CSS is ready but the dragging part is broken, @veekeys will investigate tomorrow.

File: src/modal/modal.scss (hunk 5 above)
Comment (FlyersPh9) [reply src/modal/modal.scss]: Because the parent (backdrop) has `pointer-events: none` we need to undo that for the actual dialog itself.
```scss
.iui-dialog-backdrop {
	pointer-events: none;

	.iui-dialog {
		pointer-events: initial;
	}
}
```

**Comment by FlyersPh9** [comment]
> Why we now have dialog and modal with the same CSS?

We changed the approach, should all be cleared up now.

File: src/dialog/dialog.scss (hunk 6 above)
Comment (FlyersPh9) [reply src/dialog/dialog.scss:118]: In case you wanted to re-use this mixin to get the modal styling without the animation?  I can't think of any other benefit.

File: src/modal/modal.scss [hunk 7]
```diff
@@ -2,154 +2,194 @@
...
+  &.iui-modal-visible .iui-modal {
     transform: translateY(0);
     @media (prefers-reduced-motion: no-preference) {
       transition: transform $iui-full-page-modal-slide-in-speed ease-out;
     }
   }
-  // #endregion Full Page Modal Animations for CSS
+}
+
+@mixin iui-modal-resizable {
```
[the remark this answers is not in the document]
Comment (FlyersPh9) [reply src/modal/modal.scss]: Leah, Vykintas, and I all tried this and were encountering issues.

File: backstop/scenarios/modal.js [hunk 8]
```diff
@@ -6,13 +6,22 @@ module.exports = [
     // when "Open modal" sometimes is and sometimes is not hovered.
```
[the remark this answers is not in the document]
Comment (FlyersPh9) [reply backstop/scenarios/modal.js]: I removed that & ran the test 5 times in a row, failure free.  I guess it is no longer needed?

**Review by gretanausedaite — COMMENTED** [review]
Is there a way to clean minified folder? Modal.html is still there.

File: src/modal/modal.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply src/modal/modal.scss]: yeah.. so @FlyersPh9 proved me that we better use backdrop to limit the place, where dialog can appear.. I wanted to have it without backdrop, but I guess it makes sense that way :/

**Comment by veekeys** [comment]
I have added some custom js using `translate` to position dialog. It is better than using `top, left` becuase it cares about elements size. 
My code fails, if you stop dragging and try again.. (works on initial drag, till you release the mouse).

File: src/dialog/dialog.scss (hunk 6 above)
Comment (bentleyvk) [reply src/dialog/dialog.scss:118]: But do we want to do that?

File: src/dialog/dialog.scss [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    @media screen and (max-width: 400px) {
+      max-width: 95%;
+      width: 95%;
+      min-width: 95%;
+    }
+  }
+}
+
+@mixin iui-dialog-full-page {
+  $iui-full-page-dialog-slide-in-speed: 0.3s;
```
[the remark this answers is not in the document]
Comment (FlyersPh9) [reply src/dialog/dialog.scss]: Removed these and using our `$iui-speed-fast` and `$iui-speed` instead.

File: src/dialog/dialog.scss [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    @media (prefers-reduced-motion: no-preference) {
+      transition: transform $iui-speed ease-out;
+    }
+  }
+}
+
+@mixin iui-dialog-draggable {
+  background-color: transparent;
+  pointer-events: none;
+  z-index: 998;
```
Comment (bentleyvk) [inline src/dialog/dialog.scss:97]: Why need to change it?

File: src/dialog/dialog.scss [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+  &.iui-dialog-visible .iui-dialog {
+    transform: translateY(0);
+    @media (prefers-reduced-motion: no-preference) {
+      transition: transform $iui-speed ease-out;
+    }
+  }
+}
+
+@mixin iui-dialog-draggable {
```
Comment (bentleyvk) [inline src/dialog/dialog.scss:94]: We might want to change to `position: absolute` to avoid dialogs going out of containers e.g. having app with a viewer:
image
image
Because otherwise dialog would go on sidenav and header.

File: src/dialog/dialog.scss [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+@mixin iui-dialog-content {
+  flex-grow: 2;
+  margin: 0 -#{$iui-m};
+  padding: 0 $iui-m;
+  overflow-y: auto;
+  overflow-y: overlay;
+
+  @at-root .iui-dialog-draggable & {
```
Comment (bentleyvk) [inline src/dialog/dialog.scss:178]: Why it is different for draggable?

File: src/dialog/dialog.scss (hunk 11 above)
Comment (veekeys) [reply src/dialog/dialog.scss:94]: there is position absolute on `iui-dialog`

Comment (bentleyvk) [reply src/dialog/dialog.scss:94]: But it isn't helping, I guess because parent backdrop is placed on the whole screen with `position: fixed`

File: src/dialog/dialog.scss (hunk 12 above)
Comment (FlyersPh9) [reply src/dialog/dialog.scss:178]: When draggable we change a few things:
- Cursor change to `grab`.
- The title bar area gets a bar area to grab from.
- `user-select: none` so they don't accidentally highlight text when trying to drag.

**Comment by veekeys** [comment]
Last thing left: there is a bug on initial drag.. seems like miscalc and not taking into consideration current translate value or something..

**Comment by bentleyvk** [comment]
> Last thing left: there is a bug on initial drag.. seems like miscalc and not taking into consideration current translate value or something..

I tried to figure it out but couldn't. This is just a demo for us, so for me it doesn't feel like a big deal.

File: src/dialog/dialog.scss [hunk 13]
```diff
@@ -29,6 +29,11 @@
   }
 }
 
+@mixin iui-dialog-backdrop-relative {
```
Comment (bentleyvk) [inline src/dialog/dialog.scss:32]: Why it can't go by default to draggable dialog?

**Comment by bentleyvk** [comment]
> > Last thing left: there is a bug on initial drag.. seems like miscalc and not taking into consideration current translate value or something..
> 
> I tried to figure it out but couldn't. This is just a demo for us, so for me it doesn't feel like a big deal.

Fixed!

File: src/dialog/dialog.scss (hunk 13 above)
Comment (bentleyvk) [reply src/dialog/dialog.scss:32]: But how react api will look like?

File: src/dialog/dialog.scss (hunk 12 above)
Comment (bentleyvk) [reply src/dialog/dialog.scss:185]: No, I mean these two rules. I understand that draggable header is styled differently 😀

File: src/dialog/dialog.scss (hunk 13 above)
Comment (veekeys) [reply src/dialog/dialog.scss:32]: option to choose either relative to page or relative to container?

**Review by veekeys — APPROVED** [review]

**Review by gretanausedaite — APPROVED** [review]
Looks awesome!

**Review by bentleyvk — APPROVED** [review]

File: src/dialog/dialog.scss (hunk 12 above)
Comment (FlyersPh9) [reply src/dialog/dialog.scss:185]: Was able to remove 1 line of duplicate CSS.  We discussed the rest in a chat, should be good to go now.
